mantidproject/mantid
# Add support for inelastic instruments to Discus multiple scattering correction

**Description of work.**

The DiscusMultipleScatteringCorrection previously only supported elastic scattering. This PR adds support for inelastic scattering following the approach in the original Discus code (which was designed for inelastic instruments). It includes support for both direct and indirect geometry instruments

The main adjustments to the code are:
- the structure factor is now a 2D function S(Q,w) instead of a 1D function S(Q). This has involved swapping the data type of the S(Q) variable from a histogram to a workspace in various places
- the logic to manage the new track properties after each scatter now generates a new k value as well as a new angle
- the elastic version of the code now expects an input workspace with units of momentum (k) instead of wavelength. While making changes to the input units to support inelastic it seemed to worth making this change (which has been desirable for a while) at the same time

**To test:**

Unzip the following zip file [DiscusTestFiles31200.zip](https://github.com/mantidproject/mantid/files/8187483/DiscusTestFiles31200.zip)
Update the path variable in the script below to point to where you unzipped the files
Run the script which carries out a calculation on some IRIS runs with a water sample (best to do on a Release build so it doesn't take too long)
Check it completes and generate a 3D surface plot for the Scatter_1 and Scatter_2 workspaces and check they contain data:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

path = 'c:\pathtofiles'

input_workspace = Load(os.path.join(path,'iris26176_graphite002_red.nxs'))
sqw = Load(os.path.join(path,'iris26176_graphite002_sqw.nxs'))

SetSample(InputWorkspace=input_workspace,
          Geometry={'Shape': 'Cylinder', 'Height': 4.0,'Radius': 1.0, 'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'H2-O', 'MassDensity': 1.0})

# uncomment these lines to force the calculation to be run in Direct mode. Code checks sample logs first for emode    
#input_workspace.run().addProperty("deltaE-mode", "Direct", True);
#input_workspace.run().addProperty("Ei", 1.845, True);
                                  
DiscusMultipleScatteringCorrection(InputWorkspace=input_workspace, StructureFactorWorkspace=sqw,
                                   OutputWorkspace="MuscatResults", NeutronPathsSingle=1000,
                                   NeutronPathsMultiple=1000)
```

Fixes #31200.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
